compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/codegeneration/BackendCodeGenerator.kt

```diff
@@ -319,35 +313,47 @@ fun ProgramNode.compileToKotlin(
     mainFunctionBuilder.addParameter("host", Host::class)
     mainFunctionBuilder.addParameter("runtime", Runtime::class)
 
-    // TODO - figure out right way to get unique function names here
-    val hostFunNameMap: Map<Host, String> = this.hosts.associateWith { it.name + "function" }
-
-    // create a function for each host to run
-    for (entry in hostFunNameMap) {
+    // generate code for each host
+    for (host in this.hosts) {
         val curGenerator = BackendCodeGenerator(
             this,
-            entry.key,
+            host,
             codeGenerators,
             protocolComposer
         )
 
-        objectBuilder.addFunction(
-            curGenerator.generateHostFunction(
-                entry.key,
-                entry.value,
-                mainBody
-            )
+        val classBuilder = TypeSpec.classBuilder(
```

Comment:
Looking at this again, all this logic can be moved into `BackendCodeGenerator`. I.e., add a `generateClass` function that does all this and returns a `TypeSpec`. You can make `generate` and `generateFunction` private.